### PR TITLE
Lab3

### LINKED ISSUE #565: Lab3 #315
Fix multiple commits

### PR DESCRIPTION
**Pull request by hsy3418**
Fixes #565

Review of colleague's PR #506

#### Changes proposed in this PR: 

- Create a function that sort letters and prints frequency.

### CODE REVIEW

File: 03_letters/hsy3418/main.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	return m
+}
+
+//sorLetters is a function returns a sorted slice of strings with elements
+func sortLetters(m map[rune]int) []string {
+	sortSlices := make(ByLetter, 0, len(m))
+	for key, val := range m {
+		sortSlices = append(sortSlices, fmt.Sprintf("%c:%d", key, val))
+	}
+	sort.Sort(sortSlices)
```
Comment (alfredxiao) [inline 03_letters/hsy3418/main.go]: `sort.Strings()` can be used to sort string slice. If `ByLetter` is created just for sorting, probably no need for creating this custom type.

File: 03_letters/hsy3418/main_test.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	{input: map[rune]int{'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1, 'g': 1, 'h': 1, 'i': 1, 'j': 1, 'k': 1},
+		want: []string{"a:1", "b:1", "c:1", "d:1", "e:1", "f:1", "g:1", "h:1", "i:1", "j:1", "k:1"}},
+	{input: map[rune]int{'8': 1, '9': 1, '4': 2, '3': 1, '1': 1, 'k': 2, 'd': 1},
+		want: []string{"1:1", "3:1", "4:2", "8:1", "9:1", "d:1", "k:2"}},
+}
+
+func TestLetters(t *testing.T) {
+	for _, test := range letterTests {
+		actual := letters(test.input)
+		require.Equal(t, test.want, actual)
```
Comment (alfredxiao) [inline 03_letters/hsy3418/main_test.go]: Why use require which will ignore the rest test cases when one test case fails? Why not consider using assert from github.com/stretchr/testify/assert

**Comment by davidbroughsmyth** [comment]
The PR requires an issue to be created and added to your PR comment, like this:

> Fixes #
> 
> Review of colleague's PR #
> 
> Changes proposed in this PR:
> Implement Lab1 Fibonacci numbers
> Test Lab1

See https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments

**Comment by davidbroughsmyth** [comment]
One thing the code owners will ask to reduce your commits to 1 or 2 so that you don't merge so many fix commits. You can squash several existing commits into one with git rebase -i COMMIT-HASH and git push -f.

File: 03_letters/hsy3418/main.go (hunk 1 above)
Comment (juliaogris) [reply 03_letters/hsy3418/main.go]: Please fix.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please read through the [Collected Code Review Comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments)
- [Fix your pr description](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#pr-description)
- [Limit the number of commits to one or two](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#limit-the-number-of-commits-to-one-or-two)
- [Fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages)

File: 03_letters/hsy3418/main_test.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+var sortletterTests = []struct {
+	input map[rune]int
+	want  []string
+}{
+	{input: map[rune]int{'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1, 'g': 1, 'h': 1, 'i': 1, 'j': 1, 'k': 1},
+		want: []string{"a:1", "b:1", "c:1", "d:1", "e:1", "f:1", "g:1", "h:1", "i:1", "j:1", "k:1"}},
+	{input: map[rune]int{'8': 1, '9': 1, '4': 2, '3': 1, '1': 1, 'k': 2, 'd': 1},
+		want: []string{"1:1", "3:1", "4:2", "8:1", "9:1", "d:1", "k:2"}},
```
Comment (juliaogris) [inline 03_letters/hsy3418/main_test.go:32]: Please add some non-ascii unicode test cases.

File: 03_letters/hsy3418/main.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"sort"
+	"strings"
+)
+
+var out io.Writer = os.Stdout
+
+//letters is a function returns a mapping of each letter to its frequency.
+func letters(s string) map[rune]int {
+	m := make(map[rune]int)
+	for _, runeValue := range s {
```
Comment (juliaogris) [inline 03_letters/hsy3418/main.go]: nit: `rune` is enough.

**Review by juliaogris — DISMISSED** [review]
LGTM

I've fixed the Markdown formatting of your PR description, check out the history under the timestamps. It might be worthwhile to read up on Github Flavoured MarkDown:
* https://guides.github.com/features/mastering-markdown/
* https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet

**Review by rokane — APPROVED** [review]
LGTM :) I like the effort put into including documentation for your functions as well! Great work 👍

**Pull request merged**